[Server] codegen-server support for restXml (S3)

guymguym opened the pull request:
Signed-off-by: Guy Margalit <guymguym@gmail.com>

## Motivation and Context

See #903

The motivation is to implement the server protocol code for the s3d-rs crate - still draft code in the project but the idea is to provide an S3 gateway for edge devices that handles disconnections and caching transparently. See: https://github.com/s3d-rs/s3d 

## Description

This PR adds generated server code for restXml protocol types as mentioned by @crisidev in the related issue.

Please keep in mind that I am quite new to this project and to kotlin, and I would be happy to get your review and fix anything that you may suggest.

## Testing

Added the S3 model namespace to the "blocklist" in ServerProtocolTestGenerator.kt because the server protocol is not fully implemented.

## Checklist

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

rcoh commented:
you may want to install `pre-commit`, then run `pre-commit install` to pick up our pre-commit hooks (including Kotlin autoformatters)

guymguym commented:
thanks @rcoh - updated the PR with pre-commit

guymguym commented on rust-runtime/aws-smithy-http-server/src/protocols.rs:
```diff
@@ -7,8 +7,8 @@
 
-/// Validate that the request had the standard JSON content-type header.
-pub fn check_json_content_type<B>(req: &RequestParts<B>) -> Result<(), ContentTypeRejection> {
+/// Validate that the request had the standard JSON/XML content-type header.
+pub fn check_content_type<B>(req: &RequestParts<B>) -> Result<(), ContentTypeRejection> {
```
I wasn't sure what to do here - I changed it to `check_content_type` and allow json or xml, but the error is `MissingJsonContentType`. Should I change it somehow?

crisidev replied:
The original idea was to have multiple `check_<type>_content_type()` functions and to abstract the call to it inside the protocol implementation.

We are at a very early stage of development for the server codegen and we are working to finish the support of [restJson1](https://awslabs.github.io/smithy/1.0/spec/aws/aws-restjson1-protocol.html). Because of this part of the codebase is not as generic as I would like it to be.

What can be done is to have a switch matching the protocol type inside `ServerHttpProtocolGenerator` and depending on the protocol it will produce the right `check_<type>_content_type()` identifier:

```kotlin
fun getContentTypeCheck(): String {
      when (codegenContext.protocol) {
        RestJson1Trait.ID -> {
            return "check_json_content_type"
        }
        RestXmlTrait.ID -> {
            return "check_xml_content_type"
        }
        else -> {
            TODO("Protocol ${codegenContext.protocol} not supported yet")
        }
    }
}    
```

Later on

```kotlin
...
    val contentTypeCheck = getContentTypeCheck()
    rustTemplate(
        """
           let body = request.take_body().ok_or(#{SmithyHttpServer}::rejection::BodyAlreadyExtracted)?;
           let bytes = #{Hyper}::body::to_bytes(body).await?;
           if !bytes.is_empty() {
               #{SmithyHttpServer}::protocols::${contenTypeCheck}(request)?;
               input = #{parser}(bytes.as_ref(), input)?;
           }
        """,
        *codegenScope,
        "parser" to parser,
    )
...
```

I am suggesting this to unblock your PR. I have opened an issue #987 to further extract this logic inside the Protocol logic that can be addressed later.

Notes: check_json_content_type() is called [here](https://github.com/awslabs/smithy-rs/blob/main/codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/protocols/ServerHttpProtocolGenerator.kt#L522)

guymguym replied:
I just added this to the PR - I hope this makes sense..

guymguym commented on codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/protocols/serialize/XmlBindingTraitSerializerGenerator.kt:
```diff
@@ -191,11 +192,67 @@ class XmlBindingTraitSerializerGenerator(
+                        """
+                        let mut writer = #{XmlWriter}::new(&mut out);
+                        ##[allow(unused_mut)]
+                        let mut root = writer.start_el("Error")${errorShape.xmlNamespace(root = true).apply()};
```
Is there a better way I should use to write the `"Error"`tag or it makes sense to hard-quote it here in the code template?

crisidev replied:
Probably someone working on the client codegen can answer to this..

guymguym replied:
Should we tag someone to ask for review?

crisidev replied:
@rcoh do you mind to help us here?

rcoh replied:
hardcoding `Error` appears to be correct here: https://awslabs.github.io/smithy/1.0/spec/aws/aws-restxml-protocol.html#error-response-serialization.

Note that you'll also need to handle the conditional error wrapping here. On the client side, we handle it by injecting a different error extractor but on the server side, I think you'll need to handle it slightly differently. The most parallel way would be to have a function that created an `ErrorWriter` that could be swapped as a runtime type, but I'll leave that to @crisidev to decide.

https://github.com/awslabs/smithy-rs/blob/1fa0ecfccef7e73a2886937cc1bb7935c4bf4ec0/codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/protocols/RestXml.kt#L63-L66

crisidev replied:
In the server there is no need for the injection. We handle an error coming from here as a [serialization rejections](https://github.com/awslabs/smithy-rs/blob/main/rust-runtime/aws-smithy-http-server/src/rejection.rs#L29-L34).

guymguym commented on codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/protocols/serialize/XmlBindingTraitSerializerGenerator.kt:
```diff
@@ -191,11 +192,67 @@ class XmlBindingTraitSerializerGenerator(
     override fun serverErrorSerializer(shape: ShapeId): RuntimeType {
-        TODO("Not yet implemented")
+        val errorShape = model.expectShape(shape, StructureShape::class.java)
+        val xmlMembers = httpBindingResolver.errorResponseBindings(shape)
+            .filter { it.location == HttpLocation.DOCUMENT }
+            .map { it.member }
```
I have no idea if this is correct

guymguym replied:
thanks, reesolving

guymguym commented on codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/protocols/ServerHttpProtocolGenerator.kt:
```diff
@@ -869,7 +870,7 @@ private class ServerHttpProtocolImplGenerator(
     // TODO These functions can be replaced with the ones in https://docs.rs/aws-smithy-types/latest/aws_smithy_types/primitive/trait.Parse.html
     private fun generateParseStrAsPrimitiveFn(binding: HttpBindingDescriptor): RuntimeType {
-        val output = symbolProvider.toSymbol(binding.member)
+        val output = symbolProvider.toSymbol(binding.member).makeOptional()
```
I had to fix that, but I hope this is also ok for client codegen too

crisidev replied:
I am not sure about the implications of this, I will have to check with the folks working on client codegen.

guymguym replied:
For context this change was needed to set the value in operation_deser for primitive types like bool/i32/etc 
For example this generated code:
```
                input = input.set_fetch_owner(
                    crate::operation_deser::parse_str_list_objects_v2_input_fetch_owner(v)?,
                );
```
did not compile as the setter expected an Option<bool>.
I think this is actually a must have change anyhow because the generated body always returns Ok(Some(value)) and just the return type of the function was not matching.
See here - https://github.com/awslabs/smithy-rs/blob/6325c8a277225e786920d42df935371d27b11fc4/codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/protocols/ServerHttpProtocolGenerator.kt#L883

crisidev replied:
@rcoh can you help us here?

crisidev replied:
I think we can go with makeOptional() as this stage. Input validation is on the road map and it is something we will address later.

guymguym commented on codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/protocol/ServerProtocolTestGenerator.kt:
```diff
@@ -209,7 +209,7 @@ class ServerProtocolTestGenerator(
         }
         httpRequestTestCase.host.orNull()?.also {
-            rust("""todo!("endpoint trait not supported yet");""")
+            rust("""/* todo!("endpoint trait not supported yet"); */""")
```
I am not sure yet what is needed here so any directions would be useful

crisidev replied:
Since we are at an early phase of development, we do not implement all Smithy protocol traits yet and we want this to hard fail because if a model requires functionalities that are not implemented, it should fail when it run.

guymguym replied:
Ok so this is also not failing now that I added s3 to the block-list as suggested.

guymguym replied:
So I reverted my change here.

crisidev commented:
The CI is failing. It is important that `./gradlew :codegen-server-test:test` succeeds in your local build.

It is likely some of the protocol tests we generate are not working for restXml or for the S3 model you are using. I suggest you to add the S3 model namespace to the blocklist defined in the protocol tests. You can add another check to [the if condition](https://github.com/awslabs/smithy-rs/blob/6325c8a277225e786920d42df935371d27b11fc4/codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/protocol/ServerProtocolTestGenerator.kt#L153-L158). This is, again, because we are still working on finishing the implementation of our first protocol 😃

crisidev reviewed (changes requested):
This is wonderful work! Thanks a million. I left a couple of comments and hopefully explanations to your questions and to how to move forward.

guymguym commented:
Hi @crisidev 
So I would like to ask for some help with the current failures in the test, below is what I got locally.
The first seem simple that we don't get the <?xml tag. The other are because the bucket parsing doesn't work for some reason. Any ideas of where to look would help a lot. Thanks

```
failures:

---- operation::server_get_bucket_location_test::get_bucket_location_unwrapped_output_response stdout ----
thread 'operation::server_get_bucket_location_test::get_bucket_location_unwrapped_output_response' panicked at 'assertion failed: `(left == right)`
  left: `"<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<LocationConstraint xmlns=\"http://s3.amazonaws.com/doc/2006-03-01/\">us-west-2</LocationConstraint>"`,
 right: `b"<LocationConstraint xmlns=\"http://s3.amazonaws.com/doc/2006-03-01/\"></LocationConstraint>"`', s3/rust-server-codegen/src/operation.rs:85:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

---- operation::server_list_objects_v2_test::s3_virtual_host_dualstack_accelerate_addressing_request stdout ----
thread 'operation::server_list_objects_v2_test::s3_virtual_host_dualstack_accelerate_addressing_request' panicked at 'assertion failed: `(left == right)`
  left: `ListObjectsV2Input { bucket: None, delimiter: None, encoding_type: None, max_keys: 0, prefix: None, continuation_token: None, fetch_owner: false, start_after: None, request_payer: None, expected_bucket_owner: None }`,
 right: `ListObjectsV2Input { bucket: Some("mybucket"), delimiter: None, encoding_type: None, max_keys: 0, prefix: None, continuation_token: None, fetch_owner: false, start_after: None, request_payer: None, expected_bucket_owner: None }`', s3/rust-server-codegen/src/operation.rs:413:9

---- operation::server_list_objects_v2_test::s3_virtual_host_dualstack_addressing_request stdout ----
thread 'operation::server_list_objects_v2_test::s3_virtual_host_dualstack_addressing_request' panicked at 'assertion failed: `(left == right)`
  left: `ListObjectsV2Input { bucket: None, delimiter: None, encoding_type: None, max_keys: 0, prefix: None, continuation_token: None, fetch_owner: false, start_after: None, request_payer: None, expected_bucket_owner: None }`,
 right: `ListObjectsV2Input { bucket: Some("mybucket"), delimiter: None, encoding_type: None, max_keys: 0, prefix: None, continuation_token: None, fetch_owner: false, start_after: None, request_payer: None, expected_bucket_owner: None }`', s3/rust-server-codegen/src/operation.rs:333:9

---- operation::server_list_objects_v2_test::s3_default_addressing_request stdout ----
thread 'operation::server_list_objects_v2_test::s3_default_addressing_request' panicked at 'assertion failed: `(left == right)`
  left: `ListObjectsV2Input { bucket: None, delimiter: None, encoding_type: None, max_keys: 0, prefix: None, continuation_token: None, fetch_owner: false, start_after: None, request_payer: None, expected_bucket_owner: None }`,
 right: `ListObjectsV2Input { bucket: Some("mybucket"), delimiter: None, encoding_type: None, max_keys: 0, prefix: None, continuation_token: None, fetch_owner: false, start_after: None, request_payer: None, expected_bucket_owner: None }`', s3/rust-server-codegen/src/operation.rs:215:9

---- operation::server_list_objects_v2_test::s3_operation_addressing_preferred_request stdout ----
thread 'operation::server_list_objects_v2_test::s3_operation_addressing_preferred_request' panicked at 'assertion failed: `(left == right)`
  left: `ListObjectsV2Input { bucket: None, delimiter: None, encoding_type: None, max_keys: 0, prefix: None, continuation_token: None, fetch_owner: false, start_after: None, request_payer: None, expected_bucket_owner: None }`,
 right: `ListObjectsV2Input { bucket: Some("mybucket"), delimiter: None, encoding_type: None, max_keys: 0, prefix: None, continuation_token: None, fetch_owner: false, start_after: None, request_payer: None, expected_bucket_owner: None }`', s3/rust-server-codegen/src/operation.rs:453:9

---- operation::server_list_objects_v2_test::s3_virtual_host_addressing_request stdout ----
thread 'operation::server_list_objects_v2_test::s3_virtual_host_addressing_request' panicked at 'assertion failed: `(left == right)`
  left: `ListObjectsV2Input { bucket: None, delimiter: None, encoding_type: None, max_keys: 0, prefix: None, continuation_token: None, fetch_owner: false, start_after: None, request_payer: None, expected_bucket_owner: None }`,
 right: `ListObjectsV2Input { bucket: Some("mybucket"), delimiter: None, encoding_type: None, max_keys: 0, prefix: None, continuation_token: None, fetch_owner: false, start_after: None, request_payer: None, expected_bucket_owner: None }`', s3/rust-server-codegen/src/operation.rs:254:9

---- operation::server_list_objects_v2_test::s3_virtual_host_accelerate_addressing_request stdout ----
thread 'operation::server_list_objects_v2_test::s3_virtual_host_accelerate_addressing_request' panicked at 'assertion failed: `(left == right)`
  left: `ListObjectsV2Input { bucket: None, delimiter: None, encoding_type: None, max_keys: 0, prefix: None, continuation_token: None, fetch_owner: false, start_after: None, request_payer: None, expected_bucket_owner: None }`,
 right: `ListObjectsV2Input { bucket: Some("mybucket"), delimiter: None, encoding_type: None, max_keys: 0, prefix: None, continuation_token: None, fetch_owner: false, start_after: None, request_payer: None, expected_bucket_owner: None }`', s3/rust-server-codegen/src/operation.rs:373:9


failures:
    operation::server_get_bucket_location_test::get_bucket_location_unwrapped_output_response
    operation::server_list_objects_v2_test::s3_default_addressing_request
    operation::server_list_objects_v2_test::s3_operation_addressing_preferred_request
    operation::server_list_objects_v2_test::s3_virtual_host_accelerate_addressing_request
    operation::server_list_objects_v2_test::s3_virtual_host_addressing_request
    operation::server_list_objects_v2_test::s3_virtual_host_dualstack_accelerate_addressing_request
    operation::server_list_objects_v2_test::s3_virtual_host_dualstack_addressing_request

test result: FAILED. 1 passed; 7 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

```

guymguym commented on codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/protocols/ServerHttpProtocolGenerator.kt:
```diff
@@ -519,7 +520,7 @@ private class ServerHttpProtocolImplGenerator(
                 let bytes = #{Hyper}::body::to_bytes(body).await?;
                 if !bytes.is_empty() {
-                    #{SmithyHttpServer}::protocols::check_json_content_type(request)?;
+                    #{SmithyHttpServer}::protocols::check_content_type(request)?;
```
done.

guymguym commented:
Oh ok so after I added the S3 model namespace to the blocklist defined in the protocol tests, the tests do not fail.

guymguym commented:
@crisidev I addressed the comments and the tests pass now locally.
There are still a couple of open questions on the changes that might affect the client codegen, that I could use your help to tag someone to review.
Thanks!

crisidev commented on codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/protocol/ServerProtocolTestGenerator.kt:
```diff
@@ -153,7 +153,9 @@ class ServerProtocolTestGenerator(
         // TODO: this allows to check-in RestJson protocol tests without
         // failures as the protocol is not fully implemented yet.
         // Remove it once the protocol is fully implemented.
```
Please also update the comment.

guymguym replied:
done

crisidev reviewed (approved):
LGTM. I will work on some refactoring after this is merged to close #987 and #988.

rcoh reviewed (commented):
overall changes look good to me! The real work will be getting all the protocol tests happy. Note that S3 generates an extended error id, I'm not sure if this is required for protocol compliance or not: https://github.com/awslabs/smithy-rs/blob/1aa59693eed10713dec0f3774a8a25ca271dbf39/aws/rust-runtime/aws-inlineable/src/s3_errors.rs#L32

rcoh commented:
(and for your own context, the diff and doc preview tools unfortunately rely on this being a PR directly to the repo which is why they're failing We hopefully will get those setup as something standalone at some point)

rcoh reviewed (approved):
changes to `codegen` look fine, delegating final approval to @crisidev

rcoh commented on codegen-server-test/build.gradle.kts:
```diff
@@ -25,7 +25,8 @@ data class CodegenTest(val service: String, val module: String, val extraConfig:
     CodegenTest("com.amazonaws.simple#SimpleService", "simple"),
     CodegenTest("aws.protocoltests.restjson#RestJson", "rest_json"),
-    CodegenTest("com.amazonaws.ebs#Ebs", "ebs")
+    CodegenTest("com.amazonaws.ebs#Ebs", "ebs"),
```
probably not in scope for this PR, but we should add the restXml protocol tests here:https://github.com/awslabs/smithy-rs/blob/cf46613360c184c5a45b7299858ebc72b9b71013/codegen-test/build.gradle.kts#L35-L38

but I suspect it might be a bit rough until we iron out all of the issues 😅

crisidev replied:
It in plans. We will probably work on this next week.

guymguym commented:
should I rebase?

crisidev commented:
Rebase done, let's wait for the CI.

crisidev commented:
I am pretty sure there will still be a lot of things to tweak in both restJson and RestXml, but this is a great start! Thanks @guymguym for the help, we really appreciate it.

guymguym commented:
no problem @crisidev I will be happy to look for more opportunities to contribute

guymguym commented:
And thank you for the super fast responses and helpful approach. Same goes to @rcoh.
